opta/cloud/aws/state.py

```diff
@@ -0,0 +1,230 @@
+# new-module-api
+
+from __future__ import annotations
+
+import io
+from typing import TYPE_CHECKING, Any, Dict, cast
+
+from botocore.exceptions import ClientError
+
+if TYPE_CHECKING:
+    from mypy_boto3_s3.literals import BucketLocationConstraintType
+
+from opta.core.terraform2.state import StateNotFoundError, StateStore, StoreConfig
+from opta.core.terraform2.terraform_file import TerraformFile
+from opta.exceptions import UserErrors
+from opta.utils import logger
+
+from . import _client
+
+
+class S3Store(StateStore):
+    def configure_storage(self) -> None:
+        self._configure_bucket()
+        self._configure_dynamodb_table()
+        self._configure_iam()
+
+    def configure_terraform_file(self, tf: TerraformFile) -> None:
+        tf.add_backend(
+            "s3",
+            {
+                "bucket": self._bucket,
+                "key": self.config.layer_name,
+                "dynamodb_table": self._bucket,
+                "region": self._region,
+            },
+        )
+
+    def is_storage_configured(self) -> bool:
+        # TODO: Also check for the dynamodb table and iam present
+        s3 = _client.s3(self._region)
+        try:
+            s3.get_bucket_encryption(Bucket=self._bucket)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchBucket":
+                return False
+            raise e
+        return True
+
+    def read_raw(self) -> str:
+        bucket = self._bucket
+        region = self._region
+        key = self._object_key
+        logger.debug(
+            f"Found an s3 backend in bucket {bucket} and key {key}, "
+            "gonna try to download the statefile from there"
+        )
+
+        s3 = _client.s3(region)
+
+        f = io.StringIO()
+        try:
+            s3.download_fileobj(Bucket=bucket, Key=key, Fileobj=f)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "404":
+                logger.debug("Did not find terraform state file")
+                raise StateNotFoundError()
+
+            raise
+
+        contents = f.getvalue()
+
+        return contents
+
+    @property
+    def _bucket(self) -> str:
+        c = self.config
+
+        return f"opta-tf-state-{c.org_name}-{c.layer_name}"
```

Comment:
there was a recent change to add a suffix to the bucket name.
https://github.com/run-x/opta/blob/main/opta/layer.py#L603

the complexity is to support old and new names, check with @bigbitbus on this topic as well